File: amber/entryutil.py
from __future__ import annotations

from typing import Optional

from .tlv import _tlv as _t, _varint_encode as _ve, _encode_str as _es


def build_entry_begin_payload(
    *,
    entry_id: int,
    kind: int,  # 0=file, 1=dir, 2=symlink
    path: str,
    mode: Optional[int] = None,
    mtime_sec: Optional[int] = None,
    mtime_nsec: Optional[int] = None,
    atime_sec: Optional[int] = None,
    atime_nsec: Optional[int] = None,
    size: Optional[int] = None,
    file_codec: Optional[int] = None,
    chunk_size: Optional[int] = None,
    symlink_target: Optional[str] = None,
) -> bytes:
    eb = bytearray()
    eb += _t(1, _ve(entry_id))
    eb += _t(2, _ve(kind))
    eb += _t(3, _es(path))
    if mode is not None:
        eb += _t(4, _ve(mode))
    if mtime_sec is not None:
        eb += _t(5, _ve(mtime_sec) + _ve(mtime_nsec or 0))
    if atime_sec is not None:
        eb += _t(6, _ve(atime_sec) + _ve(atime_nsec or 0))
    if size is not None:
        eb += _t(7, _ve(size))
    if file_codec is not None:
        eb += _t(8, _ve(file_codec))
    if chunk_size is not None:
        eb += _t(9, _ve(chunk_size))
    if kind == 2 and symlink_target is not None:
        eb += _t(10, _es(symlink_target))
    return bytes(eb)

File: src/entryutil.rs
use crate::error::AmberResult;
use crate::tlv::{tlv, varint_encode};

pub struct EntryBeginPayload<'a> {
    pub entry_id: u64,
    pub kind: u64,
    pub path: &'a str,
    pub mode: Option<u64>,
    pub mtime_sec: Option<u64>,
    pub mtime_nsec: Option<u64>,
    pub atime_sec: Option<u64>,
    pub atime_nsec: Option<u64>,
    pub size: Option<u64>,
    pub file_codec: Option<u64>,
    pub chunk_size: Option<u64>,
    pub symlink_target: Option<&'a str>,
}

pub fn build_entry_begin_payload(spec: EntryBeginPayload<'_>) -> AmberResult<Vec<u8>> {
    let mut out = Vec::new();
    out.extend(tlv(1, &varint_encode(spec.entry_id))?);
    out.extend(tlv(2, &varint_encode(spec.kind))?);
    out.extend(tlv(3, spec.path.as_bytes())?);
    if let Some(mode) = spec.mode {
        out.extend(tlv(4, &varint_encode(mode))?);
    }
    if let Some(mtime_sec) = spec.mtime_sec {
        out.extend(tlv(
            5,
            &[
                varint_encode(mtime_sec),
                varint_encode(spec.mtime_nsec.unwrap_or(0)),
            ]
            .concat(),
        )?);
    }
    if let Some(atime_sec) = spec.atime_sec {
        out.extend(tlv(
            6,
            &[
                varint_encode(atime_sec),
                varint_encode(spec.atime_nsec.unwrap_or(0)),
            ]
            .concat(),
        )?);
    }
    if let Some(size) = spec.size {
        out.extend(tlv(7, &varint_encode(size))?);
    }
    if let Some(file_codec) = spec.file_codec {
        out.extend(tlv(8, &varint_encode(file_codec))?);
    }
    if let Some(chunk_size) = spec.chunk_size {
        out.extend(tlv(9, &varint_encode(chunk_size))?);
    }
    if spec.kind == 2 {
        if let Some(target) = spec.symlink_target {
            out.extend(tlv(10, target.as_bytes())?);
        }
    }
    Ok(out)
}
